File: packages/docs-test/tests/typescript/sdk.ts
import "dotenv/config"
import fs from "fs"
import Supermemory, { toFile } from "supermemory"

const client = new Supermemory()

async function testDocumentsCRUD() {
  console.log("\n=== Document CRUD Operations ===")

  // Create
  const doc = await client.documents.add({
    content: "Test content for TypeScript SDK - " + Date.now(),
  })
  console.log("✓ documents.add:", doc.id)

  // Read
  const fetched = await client.documents.get(doc.id)
  console.log("✓ documents.get:", fetched.id)

  // Update
  const updated = await client.documents.update(doc.id, {
    content: "Updated content - " + Date.now(),
  })
  console.log("✓ documents.update:", updated.id)

  // Wait for processing before delete
  await new Promise((r) => setTimeout(r, 10000))

  // Delete
  await client.documents.delete(doc.id)
  console.log("✓ documents.delete")
}

async function testBatchOperations() {
  console.log("\n=== Batch Operations ===")

  const batch = await client.documents.batchAdd({
    documents: [
      { content: "Batch doc 1 - " + Date.now() },
      { content: "Batch doc 2 - " + Date.now() },
    ],
  })
  console.log("✓ documents.batchAdd:", batch)
}

async function testSearch() {
  console.log("\n=== Search ===")

  const results = await client.search.execute({ q: "test content" })
  console.log("✓ search.execute:", results.results?.length ?? 0, "results")
}

async function testFileUploads() {
  console.log("\n=== File Uploads ===")

  // Using fs.createReadStream
  const testFile = "/tmp/test-doc-upload.txt"
  fs.writeFileSync(testFile, "Test file content")
  await client.documents.uploadFile({ file: fs.createReadStream(testFile) })
  console.log("✓ uploadFile with fs.createReadStream")

  // Using File API
  await client.documents.uploadFile({ file: new File(["my bytes"], "file") })
  console.log("✓ uploadFile with File API")

  // Using toFile
  await client.documents.uploadFile({ file: await toFile(Buffer.from("my bytes"), "file") })
  console.log("✓ uploadFile with toFile")

  fs.unlinkSync(testFile)
}

async function testClientConfig() {
  console.log("\n=== Client Configuration ===")

  // Retries
  const clientWithRetries = new Supermemory({ maxRetries: 0 })
  console.log("✓ maxRetries config")

  // Timeout
  const clientWithTimeout = new Supermemory({ timeout: 20 * 1000 })
  console.log("✓ timeout config")

  // Per-request options
  await client.add({ content: "Test - " + Date.now() }, { maxRetries: 5, timeout: 5000 })
  console.log("✓ per-request options")
}

async function testRawResponse() {
  console.log("\n=== Raw Response ===")

  const response = await client.documents
    .add({ content: "Test - " + Date.now() })
    .asResponse()
  console.log("✓ .asResponse() status:", response.status)

  const { data, response: raw } = await client.documents
    .add({ content: "Test - " + Date.now() })
    .withResponse()
  console.log("✓ .withResponse() data:", data.id, "status:", raw.status)
}

async function testTypes() {
  console.log("\n=== Types ===")

  const params: Supermemory.AddParams = {
    content: "Typed content",
  }
  const response: Supermemory.AddResponse = await client.add(params)
  console.log("✓ TypeScript types work:", response.id)
}

async function testErrorHandling() {
  console.log("\n=== Error Handling ===")

  const response = await client.documents
    .add({ content: "Test - " + Date.now() })
    .catch(async (err) => {
      if (err instanceof Supermemory.APIError) {
        console.log("Caught APIError:", err.status, err.name)
      } else {
        throw err
      }
      return null
    })
  console.log("✓ Error handling pattern works:", response?.id)
}

async function main() {
  console.log("TypeScript SDK Tests")
  console.log("====================")

  await testDocumentsCRUD()
  await testBatchOperations()
  await testSearch()
  await testFileUploads()
  await testClientConfig()
  await testRawResponse()
  await testTypes()
  await testErrorHandling()

  console.log("\n====================")
  console.log("✅ All TypeScript SDK tests passed!")
}

main().catch(console.error)

File: packages/docs-test/tests/python/sdk.py
import os
import time
from pathlib import Path
from dotenv import load_dotenv
import httpx
import supermemory
from supermemory import Supermemory

load_dotenv()

client = Supermemory()


def test_documents_crud():
    """Test document CRUD operations"""
    print("\n=== Document CRUD Operations ===")

    # Create
    doc = client.documents.add(content=f"Test content - {time.time()}")
    print(f"✓ documents.add: {doc.id}")

    # Read
    fetched = client.documents.get(doc.id)
    print(f"✓ documents.get: {fetched.id}")

    # Update
    updated = client.documents.update(doc.id, content=f"Updated - {time.time()}")
    print(f"✓ documents.update: {updated.id}")

    # Wait for processing
    time.sleep(10)

    # Delete
    client.documents.delete(doc.id)
    print("✓ documents.delete")


def test_batch_operations():
    """Test batch operations"""
    print("\n=== Batch Operations ===")

    batch = client.documents.batch_add(
        documents=[
            {"content": f"Batch 1 - {time.time()}"},
            {"content": f"Batch 2 - {time.time()}"},
        ]
    )
    print(f"✓ documents.batch_add: {batch}")


def test_search():
    """Test search"""
    print("\n=== Search ===")

    results = client.search.execute(q="test content")
    print(f"✓ search.execute: {len(results.results) if results.results else 0} results")


def test_file_uploads():
    """Test file uploads"""
    print("\n=== File Uploads ===")

    test_path = Path("/tmp/test-py-upload.txt")
    test_path.write_text(f"Test content {time.time()}")
    client.documents.upload_file(file=test_path)
    print("✓ documents.upload_file with Path")
    test_path.unlink()


def test_error_handling():
    """Test error handling patterns"""
    print("\n=== Error Handling ===")

    try:
        client.documents.add(content="Test content")
        print("✓ Error handling pattern works")
    except supermemory.APIConnectionError:
        print("APIConnectionError handled")
    except supermemory.RateLimitError:
        print("RateLimitError handled")
    except supermemory.APIStatusError:
        print("APIStatusError handled")


def test_client_config():
    """Test client configuration"""
    print("\n=== Client Configuration ===")

    # Retries
    client2 = Supermemory(max_retries=0)
    print("✓ max_retries config")

    # Timeout
    client3 = Supermemory(timeout=20.0)
    print("✓ timeout config")

    # Granular timeout
    client4 = Supermemory(timeout=httpx.Timeout(60.0, read=5.0, write=10.0, connect=2.0))
    print("✓ granular timeout config")

    # Per-request options
    client.with_options(max_retries=5).documents.add(content="Test content")
    print("✓ per-request options")


def test_raw_response():
    """Test raw response access"""
    print("\n=== Raw Response ===")

    response = client.documents.with_raw_response.add(content="Test content")
    print(f"✓ with_raw_response: has headers={hasattr(response, 'headers')}")

    memory = response.parse()
    print(f"✓ parse(): {memory.id}")


def test_streaming_response():
    """Test streaming response"""
    print("\n=== Streaming Response ===")

    with client.documents.with_streaming_response.add(content="Test content") as response:
        print(f"✓ with_streaming_response: has headers={hasattr(response, 'headers')}")


def test_context_manager():
    """Test context manager"""
    print("\n=== Context Manager ===")

    with Supermemory() as temp_client:
        print("✓ Context manager works")


def main():
    print("Python SDK Tests")
    print("================")

    test_documents_crud()
    test_batch_operations()
    test_search()
    test_file_uploads()
    test_error_handling()
    test_client_config()
    test_raw_response()
    test_streaming_response()
    test_context_manager()

    print("\n================")
    print("✅ All Python SDK tests passed!")


if __name__ == "__main__":
    main()
